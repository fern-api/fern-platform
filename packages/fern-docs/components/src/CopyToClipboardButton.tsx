import { useCopyToClipboard } from "@fern-ui/react-commons";
import { composeEventHandlers } from "@radix-ui/primitive";
import { Slot } from "@radix-ui/react-slot";
import { Check, Copy } from "iconoir-react";
import {
  ComponentPropsWithoutRef,
  Fragment,
  ReactNode,
  forwardRef,
  useState,
} from "react";
import { cn } from "./cn";
import { FernButton } from "./FernButton";
import { FernTooltip, FernTooltipProvider } from "./FernTooltip";

const CopyIconButton = forwardRef<
  HTMLButtonElement,
  Omit<ComponentPropsWithoutRef<typeof FernButton>, "children"> & {
    wasJustCopied?: boolean;
  }
>(({ wasJustCopied, ...props }, ref) => {
  return (
    <FernButton
      {...props}
      ref={ref}
      className={cn("fern-copy-button group", props.className)}
      icon={wasJustCopied ? <Check /> : <Copy />}
      variant="minimal"
    />
  );
});

CopyIconButton.displayName = "CopyIconButton";

export const CopyToClipboardButton = forwardRef<
  HTMLButtonElement,
  Omit<ComponentPropsWithoutRef<"button">, "content"> & {
    content?: string | (() => string | Promise<string>);
    tooltipContent?:
      | ReactNode
      | ((props: {
          copyToClipboard: undefined | (() => void);
          setTooltipOpen: (open: boolean) => void;
        }) => ReactNode);
    asChild?: boolean;
    tooltipContentAsChild?: boolean;
    delayDuration?: number;
    disableTooltipProvider?: boolean;
    disablePortal?: boolean;
  }
>((props, ref) => {
  const {
    content,
    tooltipContent,
    children,
    asChild,
    tooltipContentAsChild,
    delayDuration,
    disableTooltipProvider,
    disablePortal,
    ...otherProps
  } = props;

  const [open, setOpen] = useState(false);
  const { copyToClipboard, wasJustCopied } = useCopyToClipboard(content);
  const Comp = asChild ? Slot : "button";
  const onClick = composeEventHandlers(props.onClick, copyToClipboard, {
    checkForDefaultPrevented: true,
  });

  const Provider = disableTooltipProvider ? Fragment : FernTooltipProvider;
  return (
    <Provider>
      <FernTooltip
        content={
          wasJustCopied ? (
            "Copied!"
          ) : tooltipContent != null ? (
            typeof tooltipContent === "function" ? (
              tooltipContent({ copyToClipboard, setTooltipOpen: setOpen })
            ) : (
              tooltipContent
            )
          ) : (
            <p>{"Copy to clipboard"}</p>
          )
        }
        open={wasJustCopied || open}
        onOpenChange={setOpen}
        asChild
        contentAsChild={tooltipContentAsChild}
        delayDuration={delayDuration}
        disableHoverableContent={
          wasJustCopied || typeof tooltipContent !== "function"
        }
        disablePortal={disablePortal}
        updatePositionStrategy="always"
      >
        {!children ? (
          <CopyIconButton
            ref={ref}
            wasJustCopied={wasJustCopied}
            disabled={copyToClipboard == null}
            onClick={onClick}
            {...otherProps}
          />
        ) : (
          <Comp
            {...otherProps}
            ref={ref}
            disabled={copyToClipboard == null}
            onClick={onClick}
          >
            {children}
          </Comp>
        )}
      </FernTooltip>
    </Provider>
  );
});

CopyToClipboardButton.displayName = "CopyToClipboardButton";
