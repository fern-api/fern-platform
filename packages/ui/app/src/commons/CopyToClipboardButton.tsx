import { useCopyToClipboard } from "@fern-ui/react-commons";
import { CopyIcon } from "@radix-ui/react-icons";
import classNames from "classnames";
import { Check } from "react-feather";
import { FernButton } from "../components/FernButton";
import { FernTooltip, FernTooltipProvider } from "../components/FernTooltip";

export declare namespace CopyToClipboardButton {
    export interface Props {
        className?: string;
        content?: string | (() => string);
        testId?: string;
        children?: (onClick: (() => void) | undefined) => JSX.Element;
    }
}

export const CopyToClipboardButton: React.FC<CopyToClipboardButton.Props> = ({
    className,
    content,
    testId,
    children,
}) => {
    const { copyToClipboard, wasJustCopied } = useCopyToClipboard(content);

    if (content == null) {
        return null;
    }

    return (
        <FernTooltipProvider>
            <FernTooltip
                content={wasJustCopied ? "Copied!" : "Copy to clipboard"}
                open={wasJustCopied ? true : undefined}
            >
                {children?.(copyToClipboard) ?? (
                    <FernButton
                        className={classNames("group", className)}
                        disabled={copyToClipboard == null}
                        onClick={copyToClipboard}
                        data-testid={testId}
                        rounded={true}
                        icon={wasJustCopied ? <Check className="size-4" /> : <CopyIcon className="size-4" />}
                        variant="minimal"
                        intent={wasJustCopied ? "success" : "none"}
                    />
                )}
            </FernTooltip>
        </FernTooltipProvider>
    );
};
