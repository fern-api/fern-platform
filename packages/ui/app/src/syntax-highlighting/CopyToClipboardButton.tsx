import { useCopyToClipboard } from "@fern-ui/react-commons";
import { CopyIcon } from "@radix-ui/react-icons";
import cn from "clsx";
import { Check } from "react-feather";
import { FernButton } from "../components/FernButton";
import { FernTooltip, FernTooltipProvider } from "../components/FernTooltip";

export declare namespace CopyToClipboardButton {
    export interface Props {
        className?: string;
        content?: string | (() => string);
        testId?: string;
        children?: (onClick: ((e: React.MouseEvent) => void) | undefined) => React.ReactNode;
        onClick?: (e: React.MouseEvent) => void;
    }
}

export const CopyToClipboardButton: React.FC<CopyToClipboardButton.Props> = ({
    className,
    content,
    testId,
    children,
    onClick,
}) => {
    const { copyToClipboard, wasJustCopied } = useCopyToClipboard(content);

    if (content == null) {
        return null;
    }

    return (
        <FernTooltipProvider>
            <FernTooltip
                content={wasJustCopied ? "Copied!" : "Copy to clipboard"}
                open={wasJustCopied ? true : undefined}
            >
                {children?.((e) => {
                    onClick?.(e);
                    copyToClipboard?.();
                }) ?? (
                    <FernButton
                        className={cn("group fern-copy-button", className)}
                        disabled={copyToClipboard == null}
                        onClickCapture={(e) => {
                            onClick?.(e);
                            copyToClipboard?.();
                        }}
                        data-testid={testId}
                        rounded={true}
                        icon={wasJustCopied ? <Check className="size-4" /> : <CopyIcon className="size-4" />}
                        variant="minimal"
                        intent={wasJustCopied ? "success" : "none"}
                        disableAutomaticTooltip={true}
                    />
                )}
            </FernTooltip>
        </FernTooltipProvider>
    );
};
