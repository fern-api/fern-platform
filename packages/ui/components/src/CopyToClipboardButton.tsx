import { useCopyToClipboard } from "@fern-ui/react-commons";
import { composeEventHandlers } from "@radix-ui/primitive";
import { Slot, Slottable } from "@radix-ui/react-slot";
import { Check, Copy } from "iconoir-react";
import { ComponentPropsWithoutRef, ReactNode, forwardRef } from "react";
import { Button } from "./FernButtonV2";
import { FernTooltip, FernTooltipProvider } from "./FernTooltip";
import { cn } from "./cn";

export const CopyToClipboardButton = forwardRef<
    HTMLButtonElement,
    Omit<ComponentPropsWithoutRef<"button">, "content"> & {
        content?: string | (() => string | Promise<string>);
        tooltipContent?: ReactNode | ((props: { copyToClipboard: undefined | (() => void) }) => ReactNode);
        asChild?: boolean;
        tooltipContentAsChild?: boolean;
        size?: ComponentPropsWithoutRef<typeof Button>["size"];
        hideIcon?: boolean;
        delayDuration?: number;
    }
>(
    (
        {
            className,
            content,
            tooltipContent,
            children,
            asChild,
            tooltipContentAsChild,
            hideIcon,
            delayDuration,
            ...props
        },
        ref,
    ) => {
        const { copyToClipboard, wasJustCopied } = useCopyToClipboard(content);

        const Comp = asChild ? Slot : Button;

        return (
            <FernTooltipProvider>
                <FernTooltip
                    content={
                        wasJustCopied ? (
                            "Copied!"
                        ) : tooltipContent != null ? (
                            typeof tooltipContent === "function" ? (
                                tooltipContent({ copyToClipboard })
                            ) : (
                                tooltipContent
                            )
                        ) : (
                            <p>{"Copy to clipboard"}</p>
                        )
                    }
                    open={wasJustCopied ? true : undefined}
                    asChild
                    contentAsChild={tooltipContentAsChild}
                    delayDuration={delayDuration}
                >
                    <Comp
                        size={hideIcon ? undefined : "icon"}
                        {...props}
                        ref={ref}
                        className={cn("group fern-copy-button", className)}
                        disabled={copyToClipboard == null}
                        onClick={composeEventHandlers(props.onClick, copyToClipboard, {
                            checkForDefaultPrevented: true,
                        })}
                    >
                        {hideIcon ? null : wasJustCopied ? <Check /> : <Copy />}
                        <Slottable>{children}</Slottable>
                    </Comp>
                </FernTooltip>
            </FernTooltipProvider>
        );
    },
);

CopyToClipboardButton.displayName = "CopyToClipboardButton";
