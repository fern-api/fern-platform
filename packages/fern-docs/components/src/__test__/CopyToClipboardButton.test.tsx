/**
 * @vitest-environment jsdom
 */

Object.defineProperty(navigator, "clipboard", {
  value: {
    writeText: async () => {
      // Ignore
    },
  },
});

import { cleanup, fireEvent, render } from "@testing-library/react";
import { CopyToClipboardButton } from "../CopyToClipboardButton";

afterEach(cleanup);

describe("CopyToClipboardButton", () => {
  it("changes content after click", () => {
    const { getByTestId } = render(
      <CopyToClipboardButton data-testid="copy-btn" content="abc" />
    );

    const innerHtmlBeforeClick = getByTestId("copy-btn").innerHTML;

    fireEvent.click(getByTestId("copy-btn"));

    const innerHtmlAfterClick = getByTestId("copy-btn").innerHTML;

    expect(innerHtmlAfterClick).not.toEqual(innerHtmlBeforeClick);
  });
});
