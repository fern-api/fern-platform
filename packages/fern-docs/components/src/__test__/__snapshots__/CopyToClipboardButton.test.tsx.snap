// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CopyToClipboardButton > renders correctly 1`] = `
<button
  aria-disabled={false}
  className="fern-copy-button group fern-button minimal normal rounded square"
  data-state="closed"
  data-testid="copy-btn"
  onBlur={[Function]}
  onClick={[Function]}
  onClickCapture={[Function]}
  onFocus={[Function]}
  onPointerDown={[Function]}
  onPointerLeave={[Function]}
  onPointerMove={[Function]}
  tabIndex={0}
>
  <span
    className="fern-button-content"
  >
    <svg
      color="currentColor"
      fill="none"
      height="1.5em"
      strokeWidth={1.5}
      viewBox="0 0 24 24"
      width="1.5em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19.4 20H9.6C9.26863 20 9 19.7314 9 19.4V9.6C9 9.26863 9.26863 9 9.6 9H19.4C19.7314 9 20 9.26863 20 9.6V19.4C20 19.7314 19.7314 20 19.4 20Z"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
      <path
        d="M15 9V4.6C15 4.26863 14.7314 4 14.4 4H4.6C4.26863 4 4 4.26863 4 4.6V14.4C4 14.7314 4.26863 15 4.6 15H9"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
  </span>
</button>
`;
